File: server/test/alda/parser/duration_test.clj
(ns alda.parser.duration-test
  (:require [clojure.test :refer :all]
            [alda.parser-util :refer (parse-with-context)]))

(deftest duration-tests
  (testing "duration"
    (is (= (parse-with-context :music-data "c2")
           '((alda.lisp/note
               (alda.lisp/pitch :c)
               (alda.lisp/duration (alda.lisp/note-length 2))))))))

(deftest dot-tests
  (testing "dots"
    (is (= (parse-with-context :music-data "c2..")
           '((alda.lisp/note
               (alda.lisp/pitch :c)
               (alda.lisp/duration (alda.lisp/note-length 2 {:dots 2}))))))))

(deftest millisecond-duration-tests
  (testing "duration in milliseconds"
    (is (= (parse-with-context :music-data "c450ms")
           '((alda.lisp/note
               (alda.lisp/pitch :c)
               (alda.lisp/duration (alda.lisp/ms 450))))))))

(deftest second-duration-tests
  (testing "duration in seconds"
    (is (= (parse-with-context :music-data "c2s")
           '((alda.lisp/note
               (alda.lisp/pitch :c)
               (alda.lisp/duration (alda.lisp/ms 2000))))))))

(deftest tie-and-slur-tests
  (testing "ties"
    (testing "ties"
      (is (= (parse-with-context :music-data "c1~2~4")
             '((alda.lisp/note
                 (alda.lisp/pitch :c)
                 (alda.lisp/duration (alda.lisp/note-length 1)
                                     (alda.lisp/note-length 2)
                                     (alda.lisp/note-length 4))))))
      (is (= (parse-with-context :music-data "c500ms~350ms")
             '((alda.lisp/note
                 (alda.lisp/pitch :c)
                 (alda.lisp/duration (alda.lisp/ms 500)
                                     (alda.lisp/ms 350))))))
      (is (= (parse-with-context :music-data "c5s~4~350ms")
             '((alda.lisp/note
                 (alda.lisp/pitch :c)
                 (alda.lisp/duration (alda.lisp/ms 5000)
                                     (alda.lisp/note-length 4)
                                     (alda.lisp/ms 350)))))))
    (testing "slurs"
      (is (= (parse-with-context :music-data "c4~")
             '((alda.lisp/note
                 (alda.lisp/pitch :c)
                 (alda.lisp/duration (alda.lisp/note-length 4))
                 :slur))))
      (is (= (parse-with-context :music-data "c420ms~")
             '((alda.lisp/note
                 (alda.lisp/pitch :c)
                 (alda.lisp/duration (alda.lisp/ms 420))
                 :slur)))))))

File: client/parser/duration_test.go
package parser

import (
	"testing"

	"alda.io/client/model"
	_ "alda.io/client/testing"
)

func cNoteWithDuration(components ...model.DurationComponent) model.Note {
	return model.Note{
		Pitch:    model.LetterAndAccidentals{NoteLetter: model.C},
		Duration: model.Duration{Components: components},
	}
}

func TestDurations(t *testing.T) {
	executeParseTestCases(
		t,
		parseTestCase{
			label: "Note with integer note length",
			given: "c2",
			expect: []model.ScoreUpdate{
				cNoteWithDuration(model.NoteLength{Denominator: 2}),
			},
		},
		parseTestCase{
			label: "Note with fractional note length",
			given: "c0.5",
			expect: []model.ScoreUpdate{
				cNoteWithDuration(model.NoteLength{Denominator: 0.5}),
			},
		},
		parseTestCase{
			label: "Note with integer note length and dots",
			given: "c2..",
			expect: []model.ScoreUpdate{
				cNoteWithDuration(model.NoteLength{Denominator: 2, Dots: 2}),
			},
		},
		parseTestCase{
			label: "Note with fractional note length and dots",
			given: "c0.5..",
			expect: []model.ScoreUpdate{
				cNoteWithDuration(model.NoteLength{Denominator: 0.5, Dots: 2}),
			},
		},
		parseTestCase{
			label: "Note with duration in milliseconds",
			given: "c450ms",
			expect: []model.ScoreUpdate{
				cNoteWithDuration(model.NoteLengthMs{Quantity: 450}),
			},
		},
		parseTestCase{
			label: "Note with duration in seconds",
			given: "c2s",
			expect: []model.ScoreUpdate{
				cNoteWithDuration(model.NoteLengthMs{Quantity: 2000}),
			},
		},
		parseTestCase{
			label: "Note with tied integer note lengths",
			given: "c1~2~4",
			expect: []model.ScoreUpdate{
				cNoteWithDuration(
					model.NoteLength{Denominator: 1},
					model.NoteLength{Denominator: 2},
					model.NoteLength{Denominator: 4},
				),
			},
		},
		parseTestCase{
			label: "Note with tied integer/fractional note lengths",
			given: "c1.5~2.5~4",
			expect: []model.ScoreUpdate{
				cNoteWithDuration(
					model.NoteLength{Denominator: 1.5},
					model.NoteLength{Denominator: 2.5},
					model.NoteLength{Denominator: 4},
				),
			},
		},
		parseTestCase{
			label: "Note with tied millisecond note lengths",
			given: "c500ms~350ms",
			expect: []model.ScoreUpdate{
				cNoteWithDuration(
					model.NoteLengthMs{Quantity: 500},
					model.NoteLengthMs{Quantity: 350},
				),
			},
		},
		parseTestCase{
			label: "Note with various tied note lengths",
			given: "c5s~4~350ms~0.5",
			expect: []model.ScoreUpdate{
				cNoteWithDuration(
					model.NoteLengthMs{Quantity: 5000},
					model.NoteLength{Denominator: 4},
					model.NoteLengthMs{Quantity: 350},
					model.NoteLength{Denominator: 0.5},
				),
			},
		},
		parseTestCase{
			label: "Slurred note with implicit duration",
			given: "c~",
			expect: []model.ScoreUpdate{
				model.Note{
					Pitch:   model.LetterAndAccidentals{NoteLetter: model.C},
					Slurred: true,
				},
			},
		},
		parseTestCase{
			label: "Slurred note with integer note length",
			given: "c4~",
			expect: []model.ScoreUpdate{
				model.Note{
					Pitch: model.LetterAndAccidentals{NoteLetter: model.C},
					Duration: model.Duration{
						Components: []model.DurationComponent{
							model.NoteLength{Denominator: 4},
						},
					},
					Slurred: true,
				},
			},
		},
		parseTestCase{
			label: "Slurred note with millisecond note length",
			given: "c420ms~",
			expect: []model.ScoreUpdate{
				model.Note{
					Pitch: model.LetterAndAccidentals{NoteLetter: model.C},
					Duration: model.Duration{
						Components: []model.DurationComponent{
							model.NoteLengthMs{Quantity: 420},
						},
					},
					Slurred: true,
				},
			},
		},
	)
}
